astropy/io/ascii/ui.py

```diff
@@ -641,6 +665,13 @@ def _guess(table, read_kwargs, format, fast_reader):
             reader = get_reader(**guess_kwargs)
 
             reader.guessing = True
+
+            if table_guess_subset:
```

Comment:
I think this is the spot where I think we want something like
```
if table_guess_subset is not None and getattr(reader, "enable_guess_limit_lines"):
```
Then the `Reader` class for any class where applying a guess limit is safe should define a class attribute `enable_guess_limit_lines = True`. So this would include all the `Basic` readers but e.g. `ECSV`, `CDS`, `IPAC` would NOT have this (or set it to `False`).

I originally imagined something to opt out, but considering there could be user-defined classes out there where there is a non-trivial header and the potential for something silly like a user configuring to check only the first 100 lines and shooting themself in the foot.

One interesting case is the C-parser fast readers. From my testing and look at the code, I think they short-circuit quickly because there is a line-by-line check that the number of data columns matches the number of header names. So we might be better off leaving those out.

Reply:
So is the idea that readers that can fail fast anyway should not be included here? Note that for ECSV we need e.g. https://github.com/astropy/astropy/pull/16850#event-13930141080 to truly fail fast and not start processing all the lines.

Reply:
I am basically rescinding my above comment, but would like to leave it for the record.